src/type_checker/checker.rs

```diff
@@ -654,6 +654,41 @@ impl<B: Backend> TypeChecker<B> {
                 typed_fn_env.pop();
                 typed_fn_env.end_forloop();
             }
+            StmtKind::IteratorLoop {
+                var,
+                iterator,
+                body,
+            } => {
+                // enter a new scope
+                typed_fn_env.nest();
+
+                // make sure that the iterator expression is an iterator,
+                // for now this means that the expression should have type `Array`
+                let iterator_typ = self.compute_type(iterator, typed_fn_env)?;
+                let iterator_typ =
+                    iterator_typ.expect("Could not compute type of iterator (TODO: better error)");
+
+                // the type of the variable is the type of the items of the iterator
+                let element_type = match iterator_typ.typ {
+                    TyKind::Array(element_type, _len) => *element_type,
+                    TyKind::GenericSizedArray(element_type, _size) => *element_type,
+                    _ => {
+                        return Err(self.error(
```

Comment:
maybe add a negative test (`negative_tests.rs`) to cover this scenario, this would help remind this type check should be covered in the tast phase if it happens to refactor at later stage :D